crates/schema/src/def.rs

```diff
@@ -0,0 +1,640 @@
+
+    /// Get a reference to this definition.
+    fn key(&self) -> Self::Key<'_>;
+
+    /// Look up this entity in the module def.
+    fn lookup<'a>(module_def: &'a ModuleDef, key: Self::Key<'_>) -> Option<&'a Self>;
+}
+
+/// A data structure representing the validated definition of a database table.
+///
+/// Cannot be created directly. Instead, add a [spacetimedb_sats::db::raw_def::RawTableDef] to a [spacetimedb_sats::db::raw_def::RawDatabaseDef] and call [spacetimedb_sats::db::raw_def::RawDatabaseDef::validate].
+///
+/// This struct holds information about the table, including its name, columns, indexes,
+/// constraints, sequences, type, and access rights.
+///
+/// Validation rules:
+/// - The table name must be a valid identifier.
+/// - The table's columns must be sorted according to [crate::db::ordering::canonical_ordering].
+/// - The table's indexes, constraints, and sequences must be sorted by their keys.
+/// - The table's column types may refer only to types in the containing DatabaseDef's typespace.
+/// - The table's column names must be unique.
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[non_exhaustive]
+pub struct TableDef {
+    /// The name of the table.
+    /// Unique within a module, acts as the table's identifier.
+    /// Must be a valid [crate::db::identifier::Identifier].
+    pub name: Identifier,
+
+    /// A reference to a `ProductType` containing the columns of this table.
+    /// This is the single source of truth for the table's columns.
+    /// All elements of the `ProductType` must have names.
+    ///
+    /// Like all types in the module, this must have the [default element ordering](crate::db::default_element_ordering), UNLESS a custom ordering is declared via `ModuleDef.misc_exports` for this type.
+    pub product_type_ref: AlgebraicTypeRef,
+
+    /// The primary key of the table, if present. Must refer to a valid column.
+    ///
+    /// Currently, there must be a unique constraint and an index corresponding to the primary key.
+    /// Eventually, we may remove the requirement for an index.
+    ///
+    /// The database engine does not actually care about this, but client code generation does.
+    pub primary_key: Option<ColId>,
+
+    /// The columns of this table. This stores the information in
+    /// `product_type_ref` in a more convenient-to-access format.
+    pub columns: Vec<ColumnDef>,
+
+    /// The indices on the table, indexed by name.
+    pub indexes: IdentifierMap<IndexDef>,
+
+    /// The unique constraints on the table, indexed by name.
+    pub unique_constraints: IdentifierMap<UniqueConstraintDef>,
+
+    /// The sequences for the table, indexed by name.
+    pub sequences: IdentifierMap<SequenceDef>,
+
+    /// The schedule for the table, if present.
+    pub schedule: Option<ScheduleDef>,
+
+    /// Whether this is a system- or user-created table.
+    pub table_type: StTableType,
+
+    /// Whether this table is public or private.
+    pub table_access: StAccess,
```

Comment:
Could we change those to be enums in the new version?

Those 2 being special strings has always been awkward to work with in the SDKs. The only argument for them being strings I heard is extensibility, but 1) we can add more variants to the enums just as easily and 2) we have ModuleDef versioning now anyways.

Reply:
Uh, these are enums. Possibly they are serialized as strings?

Reply:
Yeah they are enums in Rust but their SATS implementations manually serialise them as strings.

Since it's a ModuleDef-breaking change, I figured this PR would be a good place to introduce new versions of these enums with the standard enum serialisation instead.

Reply:
Ahh, I missed that entirely. I'll make the change.

Reply:
Agreed with this by the way